server/migrations/2020-06-30-135809_remove_mat_views/up.sql

```diff
@@ -0,0 +1,576 @@
+    delete from post_aggregates_fast where id = NEW.post_id;
+    insert into post_aggregates_fast select * from post_aggregates_view where id = NEW.post_id;
+  END IF;
+
+  return null;
+end $$;
+
+drop trigger refresh_post_like on post_like;
+create trigger refresh_post_like
+after insert or update or delete
+on post_like
+for each row
+execute procedure refresh_post_like();
+
+-- comment_like
+-- select id, score, my_vote from comment_fast_view where id = 29 and user_id = 4;
+-- Sample insert 
+-- insert into comment_like(user_id, comment_id, post_id, score) values (4, 29, 51, 1);
+-- Sample delete
+-- delete from comment_like where user_id = 4 and comment_id = 29;
+-- Sample update
+-- update comment_like set score = -1 where user_id = 4 and comment_id = 29;
+create or replace function refresh_comment_like()
+returns trigger language plpgsql
+as $$
+begin
+  -- TODO possibly select from comment_fast to get previous scores, instead of re-fetching the views?
+  IF (TG_OP = 'DELETE') THEN
+    delete from comment_aggregates_fast where id = OLD.comment_id;
+    insert into comment_aggregates_fast select * from comment_aggregates_view where id = OLD.comment_id;
+  ELSIF (TG_OP = 'UPDATE') THEN
+    delete from comment_aggregates_fast where id = NEW.comment_id;
+    insert into comment_aggregates_fast select * from comment_aggregates_view where id = NEW.comment_id;
+  ELSIF (TG_OP = 'INSERT') THEN
+    delete from comment_aggregates_fast where id = NEW.comment_id;
+    insert into comment_aggregates_fast select * from comment_aggregates_view where id = NEW.comment_id;
+  END IF;
+
+  return null;
+end $$;
+
+drop trigger refresh_comment_like on comment_like;
+create trigger refresh_comment_like
+after insert or update or delete
+on comment_like
+for each row
+execute procedure refresh_comment_like();
+
+-- Community user ban
+
+drop trigger refresh_community_user_ban on community_user_ban;
+create trigger refresh_community_user_ban
+after insert or update or delete
+on community_user_ban
+for each row
+execute procedure refresh_community_user_ban();
+
+-- select creator_name, banned_from_community from comment_fast_view where user_id = 4 and content = 'test_before_ban';
+-- select creator_name, banned_from_community, community_id from comment_aggregates_fast where content = 'test_before_ban';
+-- Sample insert 
+-- insert into comment(creator_id, post_id, content) values (1198, 341, 'test_before_ban');
+-- insert into community_user_ban(community_id, user_id) values (2, 1198);
+-- Sample delete
+-- delete from community_user_ban where user_id = 1198 and community_id = 2;
+-- delete from comment where content = 'test_before_ban';
+-- update comment_aggregates_fast set banned_from_community = false where creator_id = 1198 and community_id = 2;
+create or replace function refresh_community_user_ban()
+returns trigger language plpgsql
+as $$
+begin
+  -- TODO possibly select from comment_fast to get previous scores, instead of re-fetching the views?
```

Comment:
I would say the concern is more than the ban status is kept on a "stats" table when it could live on the regular community record and if we need access to it from the fast tables just join them.  Basically my take is the stats tables could have much less data currently and just contain the keys needed to join what is otherwise acceptably normalized data

Reply:
I think I see what you're saying... basically that `post_aggregates_view` should only contain the counts type data, and not anything from simple joins with normalized data.

One issue I have with that tho, is that its a LOT, and having them in one view makes it much easier to see, example

```
create view post_aggregates_view as
select        
p.*,
(select u.banned from user_ u where p.creator_id = u.id) as banned,
(select cb.id::bool from community_user_ban cb where p.creator_id = cb.user_id and p.community_id = cb.community_id) as banned_from_community,
(select actor_id from user_ where p.creator_id = user_.id) as creator_actor_id,
(select local from user_ where p.creator_id = user_.id) as creator_local,
(select name from user_ where p.creator_id = user_.id) as creator_name,
(select avatar from user_ where p.creator_id = user_.id) as creator_avatar,
(select actor_id from community where p.community_id = community.id) as community_actor_id,
(select local from community where p.community_id = community.id) as community_local,
(select name from community where p.community_id = community.id) as community_name,
(select removed from community c where p.community_id = c.id) as community_removed,
(select deleted from community c where p.community_id = c.id) as community_deleted,
(select nsfw from community c where p.community_id = c.id) as community_nsfw,
(select count(*) from comment where comment.post_id = p.id) as number_of_comments,
coalesce(sum(pl.score), 0) as score,
count (case when pl.score = 1 then 1 else null end) as upvotes,
count (case when pl.score = -1 then 1 else null end) as downvotes,
...
```

Reply:
I would say theres nothing wrong w/ a view especially to "hide" sql logic from an interface or to make things easier to see a whole, but in practice we're just kind of shuffling the responsibility elsewhere and doing some clever processing to work around it.  So let's take the above- in theory it could be like:
```
create view post_aggregates_view as
select        
p.*,
(select u.banned from user_ u where p.creator_id = u.id) as banned,
(select cb.id::bool from community_user_ban cb where p.creator_id = cb.user_id and p.community_id = cb.community_id) as banned_from_community,
(select actor_id from user_ where p.creator_id = user_.id) as creator_actor_id,
(select local from user_ where p.creator_id = user_.id) as creator_local,
(select name from user_ where p.creator_id = user_.id) as creator_name,
(select avatar from user_ where p.creator_id = user_.id) as creator_avatar,
(select actor_id from community where p.community_id = community.id) as community_actor_id,
(select local from community where p.community_id = community.id) as community_local,
(select name from community where p.community_id = community.id) as community_name,
(select removed from community c where p.community_id = c.id) as community_removed,
(select deleted from community c where p.community_id = c.id) as community_deleted,
(select nsfw from community c where p.community_id = c.id) as community_nsfw,
-- new stuff
(select  total_comments from post_stats where post_stats.post_id = p.id) as number_of_comments,
--- ^ repeat this for each "count" or "sum"
...
```
And you can have a view which gives you the column names/logic you want while also not needing each view to read/process so much data.  The important thing going into this is that joins are fine and cheap, reading the stats of your entire DB frequently is not and right now we have them conflated (the views are being used to hide heavy processing just as much as they are providing the api a good interface).  So ideally we can get "stats" into stats tables, join everything up as needed, and then wrap the joins into a view so your api has consistency and we kinda solve everything at once by letting the views handle consistency, and letting dedicated stats tables + triggers handle your processing